librad/src/net/upgrade.rs

```diff
@@ -19,36 +19,60 @@
 //! (irrespective of ALPN, which applies per-connection). This module implements
 //! the negotiation protocol.
 
-use std::{io, marker::PhantomData, ops::Deref, pin::Pin};
+use std::{
+    fmt::{self, Debug, Display},
+    io,
+    marker::PhantomData,
+    ops::Deref,
+    pin::Pin,
+    time::Duration,
+};
 
 use futures::{
-    future::{BoxFuture, FutureExt},
-    io::{AsyncRead, AsyncWrite},
-    sink::SinkExt,
-    stream::TryStreamExt,
+    future::{self, TryFutureExt as _},
+    io::{AsyncRead, AsyncReadExt, AsyncWrite, AsyncWriteExt},
     task::{Context, Poll},
 };
-use futures_codec::Framed;
-use minicbor::{Decode, Encode};
+use futures_timer::Delay;
 use thiserror::Error;
 
-use crate::{
-    git::p2p::transport::GitStream,
-    net::{
-        codec::{CborCodec, CborCodecError},
-        quic,
-    },
-};
+use crate::git::p2p::transport::GitStream;
+
+/// Timeout waiting for an [`UpgradeRequest`].
+const RECV_UPGRADE_TIMEOUT: Duration = Duration::from_secs(23);
```

Comment:
That's _very_ specific. What's special about 23?

Reply:
It is a magical constant handed down from generations and generations of networks programmors. Also, it is very convenient to type.

Reply:
:mage:

Reply:
The truth is, I don't know what that number should be. It should wait for long enough to account for very slow links, but then also not _that_ long, and who knows what that `async` runtime is up to.

When you see me using one of `23`, `69` or `42`, it is usually a "dunno number".

Reply:
Cool :+1: Do you think it would be worth tossing that fact in for anyone else that comes across this?

Reply:
Is the upgrade timeout distinct from a normal time out to first byte? As we don't have any proof of work in the upgade I wonder how much signal is there.

Reply:
That's a good point. I don't think there is any timing out on `quinn::RecvStream` -- which does make sense: in QUIC, a stream is created by writing data to it (essentially). So, we must have received a `STREAM` frame, but not necessarily the 4 bytes we're looking for. I assume it would also be legit that this frame is empty (has no data), e.g. when it falls on the boundary of a UDP packet.

Since we have determined that we need to send keep-alives, a peer could just tar pit us by not sending anything.